### PR TITLE
Add support for starlark's unused_inputs_list to ngc-wrapped.

### PR DESCRIPTION
**Pull request by roydor**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

ngc-wrapped build actions are dependent on the full transitive closure of input files, making long cascading template compilations common when changing the public api somewhere in a dependency chain, even if those changed files are not used during compilation.

Issue Number: N/A

## What is the new behavior?
Implements support for starlark's unused_inputs_list for ngc-wrapped build operations.  This let's bazel skip targets where the only changed inputs are known to be unused, improving build times and minimizing cascading angular builds.

See:
https://docs.bazel.build/versions/main/skylark/lib/actions.html#:~:text=of%20the%20action.-,unused_inputs_list,-File%3B%20or%20None

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
I am very new to github, and ask some patience as i struggle through this pull request.

### CODE REVIEW

**Review by josephperrott — COMMENTED** [review]
This LGTM, but I want to have @devversion take a look as well

**Review by josephperrott — APPROVED** [review]
LGTM

File: packages/bazel/src/ngc-wrapped/index.ts [hunk 1]
```diff
@@ -22,7 +22,9 @@ type CompilerCliModule =
 
 // Add devmode for blaze internal
 interface BazelOptions extends ExternalBazelOptions {
+  allowedInputs?: string;
   devmode?: boolean;
+  unusedInputsListPath?: string;
```
Comment (devversion) [inline packages/bazel/src/ngc-wrapped/index.ts:27]: Are there any plans to expose the changes to `@bazel/typescript` publicly as well? Otherwise Angular externally will not be able to benefit from this change right? (or should we take care of this ourselves?)

File: packages/bazel/src/ngc-wrapped/index.ts [hunk 2]
```diff
@@ -443,9 +445,64 @@ export function compile({
...
+  const unusedInputs: string[] = [];
+  for (const f of bazelOpts.allowedInputs) {
+    // A ts/x file is unused if it was not found directly in the used sources.
+    if ((f.endsWith('.ts') || f.endsWith('.tsx')) && !usedFiles.has(f)) {
+      unusedInputs.push(f);
+      continue;
+    }
+
+    // A directory is unused if no descendants were found in the used sources.
+    if (f.endsWith('/*')) {
```
[the remark this answers is not in the document]
Comment (devversion) [reply packages/bazel/src/ngc-wrapped/index.ts:489]: Also, does this logic actually work? what happens if a nested file inside a tree artifact is used? The logic above would not add the actual tree artifact root to the `usedDirs` right? (breaking the logic)

**Review by devversion — APPROVED** [review]
LGTM overall. I have left two minor comments/questions and there is lint failing. You need to make sure that the `Merge commit` is deleted (we just need a rebase, compared to a merge)

File: packages/bazel/src/ngc-wrapped/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (roydor) [reply packages/bazel/src/ngc-wrapped/index.ts:489]: ts_library and ng_module do support tree artifacts, and in the bzl it turns it from directory notation to /* notation by adding the simple *., which is what inspired this 'reversal logic'.

That being said, if tree artifacts do allow nested levels of directories, you're correct this doesn't work.
And after following up with the bazel team, it seems tree artifacts don't work as I expected with this feature.

In summary, you're correct, this doesn't work.  We'll want to iterate the file system to figure out what to do correctly here, which is better anyways, because then you could have a partial directory marked as unused rather than all or nothing.
